### PR TITLE
Add literal extraction from array creation expressions

### PR DESCRIPTION
**Pull request by fwendland**
- analyzes array creation expressions and extracts literals from its initializer
- reworks evaluation to support more generic `MarkIntermediateResult` as value; thereby,
  allowing value to be lists
- introduce builtin `_subset` for analysis of array creation expressions in MARK
- add test case for array extraction

### CODE REVIEW

File: src/main/java/de/fraunhofer/aisec/codyze/legacy/crymlin/builtin/Subset.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.HashSet;
+
+/**
+ * Method signature: _subset(List subset, List superset)
+ * <p>
+ * This Builtin determines if the first list is a subset of the second list.
+ * <p>
+ * In case of an error, this Builtin returns an ErrorValue;
+ */
+public class Subset implements Builtin {
```
Comment (KuechA) [inline src/main/java/de/fraunhofer/aisec/codyze/legacy/crymlin/builtin/Subset.java:20]: Why don't we write this class in kotlin?

File: src/main/java/de/fraunhofer/aisec/codyze/legacy/analysis/markevaluation/ExpressionEvaluator.java [hunk 2]
```diff
@@ -275,92 +275,98 @@ private Map<Integer, MarkIntermediateResult> evaluateComparisonExpr(ComparisonEx
...
-
-						if (o != null) {
-							String inner = ExpressionHelper.toComparableString(o);
-							if (comp.compare(ExpressionHelper.toComparableString(left), inner) == 0) {
-								cv = ConstantValue.of(true);
-								cv.addResponsibleNodesFrom((ConstantValue) o);
-								break;
+			if (lr instanceof ListValue) {
+				log.error("Cannot compare against ListValue: {}", lr);
+				combinedResult.put(key, ErrorValue.newErrorValue("unsupported MARK operation"));
```
Comment (KuechA) [inline src/main/java/de/fraunhofer/aisec/codyze/legacy/analysis/markevaluation/ExpressionEvaluator.java:282]: So we can't compare two lists on equality (`==`) and unequality (`!=`)?

File: src/main/java/de/fraunhofer/aisec/codyze/legacy/analysis/markevaluation/ExpressionEvaluator.java [hunk 3]
```diff
@@ -275,92 +275,98 @@ private Map<Integer, MarkIntermediateResult> evaluateComparisonExpr(ComparisonEx
...
 			Integer key = entry.getKey();
-			ConstantValue leftBoxed = (ConstantValue) getcorrespondingLeftResult(leftResult, key);
-			Object left = leftBoxed.getValue();
-
-			if (entry.getValue() instanceof ListValue) {
-
-				if (op.equals("in")) {
-					ListValue l = (ListValue) entry.getValue();
-					ConstantValue cv = ConstantValue.of(false);
+			MarkIntermediateResult lr = getcorrespondingLeftResult(leftResult, key);
```
Comment (KuechA) [inline src/main/java/de/fraunhofer/aisec/codyze/legacy/analysis/markevaluation/ExpressionEvaluator.java:278]: The variable names are very cryptic.

File: src/main/java/de/fraunhofer/aisec/codyze/legacy/analysis/MarkIntermediateResult.java [hunk 4]
```diff
@@ -20,4 +24,6 @@ public boolean isListResult() {
 		return resultType == ResultType.LIST;
 	}
 
+	public abstract Set<Node> getResponsibleNodes();
```
Comment (KuechA) [inline src/main/java/de/fraunhofer/aisec/codyze/legacy/analysis/MarkIntermediateResult.java:27]: Javadoc missing

File: src/main/java/de/fraunhofer/aisec/codyze/legacy/crymlin/builtin/Subset.java (hunk 1 above)
Comment (fwendland) [reply src/main/java/de/fraunhofer/aisec/codyze/legacy/crymlin/builtin/Subset.java:20]: ~~'Cause the other builtins are in Java too~~ 

Let me convert it.

**Review by KuechA — APPROVED** [review]

**Pull request merged**